### DISCUSSION TITLE
How to reduce the cost of computing partial derivatives?

### DISCUSSION THREAD

**Discussion by antalszava** [jax-ml/jax#13690] [Q&A]
JAX doesn't allow marking parameters as trainable or non-trainable, but rather allows passing in the `argnums` argument to transforms performing differentiation (e.g., `jax.grad`, `jax.jacobian`, etc.). This does seem to lead to a behaviour where custom gradient rules (e.g., `custom_jvp`) need to be defined by users such that the rule entails computing the full Jacobian. This is the case even if only a subset of input arguments were selected by the user when using `argnums`, only for JAX to "choose" the specific partial derivatives requested e.g., by using the `tangents` argument (using the conventional naming of a `custom_jvp` rule).

How can one detect the input arguments wrt. which we'd like to compute partial derivatives such that the cost of computing partial derivatives is reduced?

Let's take the following example of `jax.custom_jvp` that's present in the docs:
```py
import jax
from jax import numpy as jnp
from jax import config

config.update("jax_enable_x64", True)

@jax.custom_jvp
def f(x, y):
      return jnp.sin(x) * y

@f.defjvp
def f_jvp(primals, tangents):
    x, y = primals
    x_dot, y_dot = tangents
    primal_out = f(x, y)
    print(y_dot) # <--- we inspect the tangent corresponding to the 2nd input argument that's fixed
    tangent_out = jnp.cos(x) * x_dot * y + jnp.sin(x) * y_dot
    return primal_out, tangent_out
```
We can then get the gradient of `f` via:
```py
a = jnp.array(0.3)
b = jnp.array(0.98)
print(jax.grad(f, argnums=[0])(a, b))
```
```
0.0
(Array(0.93622976, dtype=float64, weak_type=True),)
```
My understanding here is that in the line `tangent_out = jnp.cos(x) * x_dot * y + jnp.sin(x) * y_dot`, the first term of the addition (`jnp.cos(x) * x_dot * y`) is non-zero whereas the second term (`jnp.sin(x) * y_dot`) is zero. The second term is zero because `y_dot` is zero. Nonetheless, we have still computed `jnp.sin(x)` for no particular reason. Of course in this example computing `jnp.sin(x)` is not expensive to do, but `jnp.sin` can easily be replaced by another function call whose evaluation incurs significant costs due to its resource requirements (e.g., running something on dedicated hardware like HPC systems, GPU, TPU, quantum computing hardware, etc.).

What is the suggested way for defining a `custom_jvp` rule such that they are optimal in their costs when computing partial derivatives? One approach that came to mind was inspecting the tangents, however, that could make things more complex because some transform like `jax.jit` transform the input arguments. Let me know if that's the suggested way forward here though.
